File: ReadersWriterLock/Library.cs
namespace ReadersWriterLock
{
    public partial class Library : Form
    {
        private readonly ReaderWriterLock readWriteLock = new ReaderWriterLock();
        private readonly Random random = new Random();
        private bool textBoxUpdateRequested = false;

        public Library()
        {
            InitializeComponent();
        }

        private async void BtnStart_Click(object sender, EventArgs e)
        {
            TxtBox.Clear();
            textBoxUpdateRequested = true;
            List<Task> tasks = new List<Task>();

            for (int i = 1; i <= 5; i++)
            {
                Reader reader = new Reader($"Reader {i}", readWriteLock.readLock(), random.Next(200, 400), TxtBox);
                tasks.Add(Task.Run(() => reader.Run()));
            }

            for (int i = 1; i <= 2; i++)
            {
                Writer writer = new Writer($"Writer {i}", readWriteLock.writeLock(), random.Next(200, 400), TxtBox);
                tasks.Add(Task.Run(() => writer.Run()));
            }

            await Task.WhenAll(tasks);
        }

        public bool IsTextBoxUpdateRequested()
        {
            lock (this)
            {
                return textBoxUpdateRequested;
            }
        }

        public void ResetTextBoxUpdateFlag()
        {
            lock (this)
            {
                textBoxUpdateRequested = false;
            }
        }
    }
}

File: Java/ReadersWriterLockJava/src/Library.java
import javax.swing.*;
import java.awt.event.ActionEvent;
import java.util.Random;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

public class Library extends JFrame {
    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
    private final Random random = new Random();
    private boolean textBoxUpdateRequested = false;
    private JTextArea txtBox;

    public Library() {
        initComponents();
    }

    private void initComponents() {
        setTitle("Library");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setSize(400, 300);
        setLocationRelativeTo(null);

        txtBox = new JTextArea();
        JScrollPane scrollPane = new JScrollPane(txtBox);
        getContentPane().add(scrollPane);

        JButton btnStart = new JButton("Start");
        btnStart.addActionListener(this::btnStartActionPerformed);
        getContentPane().add(btnStart, "South");
    }

    private void btnStartActionPerformed(ActionEvent evt) {
        txtBox.setText("");
        textBoxUpdateRequested = true;
        for (int i = 1; i <= 5; i++) {
            Reader reader = new Reader("Reader " + i, readWriteLock.readLock(), random.nextInt(200) + 200, txtBox);
            new Thread(reader).start();
        }

        for (int i = 1; i <= 2; i++) {
            Writer writer = new Writer("Writer " + i, readWriteLock.writeLock(), random.nextInt(200) + 200, txtBox);
            new Thread(writer).start();
        }
    }

    public boolean isTextBoxUpdateRequested() {
        return textBoxUpdateRequested;
    }

    public void resetTextBoxUpdateFlag() {
        textBoxUpdateRequested = false;
    }

    public static void main(String[] args) {
        java.awt.EventQueue.invokeLater(() -> new Library().setVisible(true));
    }
}